src/java_bytecode/java_bytecode_convert_method.cpp

```diff
@@ -263,7 +263,41 @@ void java_bytecode_convert_method_lazy(
   symbol_tablet &symbol_table)
 {
   symbolt method_symbol;
-  typet member_type=java_type_from_string(m.descriptor);
+
+  typet member_type;
+  if(m.signature.has_value())
+  {
+    try
+    {
+      typet member_type_s=java_type_from_string(
+        m.signature.value(),
+        id2string(class_symbol.name));
+      INVARIANT(member_type_s.id()==ID_code, "Must be code type");
+      typet member_type_d=java_type_from_string(
+        m.descriptor,
+        id2string(class_symbol.name));
+      INVARIANT(member_type_d.id()==ID_code, "Must be code type");
+      if(to_code_type(member_type_s).parameters().size()==
+         to_code_type(member_type_d).parameters().size())
+      {
+        member_type=member_type_s;
+      }
+      else
+      {
+        // TODO: by reverting to descriptor, we are loosing generics info. In
+        // the future, we may want to extract that info from signature.
+        member_type=member_type_d;
+      }
+    }
+    catch(unsupported_java_class_signature_exceptiont &e)
+    {
+      member_type=java_type_from_string(m.descriptor);
+    }
+  }
+  else
+  {
+    member_type=java_type_from_string(m.descriptor);
+  }
```

Comment:
Is there any mileage in slightly restructuring the code so that you always initialise `member_type` to `java_type_from_string(m.descriptor)`, then do your `if(m.signature.has_value())` and `try` block, then your `catch` block would just be empty (probably with a comment to explain) and you could lose the final `else` block? That potentially adds a cost in wasted work, but looses some duplicate code and also loses the (depending on who you ask) questionable use of exceptions for control flow :-) @thk123 might have a better view than me here though.

Reply:
Another small comment having thought about this a bit overnight - whether you restructure the code or not, could you add a comment prior to, or inside, this block of code to explain the rationale for having this code? While we currently have an understanding of why we have to do this dance with the signature/descriptor, in a few weeks/months time it might not be so fresh in our memories, or if teams shuffle around and someone new to the generics stuff comes across this code, the first question they'll probably have is 'Why are we doing this?!?' :-)

Reply:
Probably the easiest way to deal with both these comments is to pull it out into a function (you can add a comment on the function and directly assign the `member_type`. I think doing the descriptor up front is a reasonable suggestion as is only wasted work when there is no signature which actually based on what Matthias said can't happen for methods even without debug symbols.